README.md
# Serverless API Response

![Build Status](https://github.com/janis-commerce/sls-api-response/workflows/Build%20Status/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/janis-commerce/sls-api-response/badge.svg?branch=master)](https://coveralls.io/github/janis-commerce/sls-api-response?branch=master)
[![npm version](https://badge.fury.io/js/%40janiscommerce%2Fsls-api-response.svg)](https://www.npmjs.com/package/@janiscommerce/sls-api-response)

A package to handle Serverless API Responses

## Installation

```
npm install @janiscommerce/sls-api-response
```

## Usage

```js
'use strict';

const { ApiResponse } = require('@janiscommerce/sls-api-response');

// Successful response (default statusCode is 200)
ApiResponse.send({
	body: {
		foo: 'bar'
	}
});

// Successful response with response headers
ApiResponse.send({
	body: {
		foo: 'bar'
	},
	headers: {
		'x-foo': 'bar'
	}
});

// Successful response with simple cookie
ApiResponse.send({
	body: {
		foo: 'bar'
	},
	cookies: {
		'my-cookie': 'bar'
	}
});

// Successful response with complex cookie
ApiResponse.send({
	body: {
		foo: 'bar'
	},
	cookies: {
		'my-cookie': {
			value: 'bar',
			httpOnly: true,
			secure: true,
			path: '/',
			expires: new Date(), // Or (new Date()).toUTCString()
			domain: '.example.com'
		}
	}
});

// Error response (default statusCode is 500)
ApiResponse.sendError(new Error('We have a problem'));
// Response body: { "message": "We have a problem" }

// Error response for a request with a client
ApiResponse.sendError(new Error('We have a problem'), 'some-client-code');
// Response body: { "message": "We have a problem" }

// Error response with custom status code
const error = new Error('You have a problem');
error.statusCode = 400;
ApiResponse.sendError(error);
// Response body: { "message": "You have a problem" }

// Error response with custom response body
const error = new Error('You have a problem');
error.body = { customErrorMessage: "Bad request body" }
ApiResponse.sendError(error);
// Response body: { "customErrorMessage": "Bad request body" }
```

## Configuration

In order to work properly, you have to set valid response templates in each function configuration. An example can be found [here](https://github.com/janis-commerce/sls-api-rest#function-minimal-configuration)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.5.0] - 2020-11-11
### Added
- Method `sendError` now accepts a client code for customizing error log

## [1.4.0] - 2020-11-10
### Added
- New method `sendError` to improve error handling

## [1.3.1] - 2020-06-23
### Fixed
- Error response with messageVariables are now stringified

## [1.3.0] - 2020-06-23
### Added
- Added support for error response with messageVariables

## [1.2.1] - 2019-10-30
### Fixed
- Removed side effect on send method

## [1.2.0] - 2019-08-05
### Added
- Response headers and cookies support

## [1.1.1] - 2019-06-25
### Fixed
- Package json, eslint and git hooks standarized
- Source code dir standarized

## [1.1.0] - 2019-06-14
### Added
- Serverless offline compatibility

## [1.0.1] - 2019-06-14
### Fixed
- README badges

## [1.0.0] - 2019-06-14
### Added
- Project inited
- Serverless API Response
- Tests


tests/api-response.js
'use strict';

const assert = require('assert');

const { ApiResponse } = require('../lib');

describe('API Response', () => {

	describe('send', () => {

		it('Should set statusCode 200 if no statusCode is specified', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				}
			};

			const { statusCode, statusCodeForPatternMatching } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(statusCode, 200);
			assert.deepStrictEqual(statusCodeForPatternMatching, '[200]');
		});

		it('Should JSON stringify the body if statusCode is not defined', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				}
			};

			const { body } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(body, JSON.stringify(originalObject.body));
		});

		it('Should JSON stringify the body if statusCode is successfull', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				statusCode: 200
			};

			const { body } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(body, JSON.stringify(originalObject.body));
		});

		it('Should set statusCodeForPatternMatching properly if statusCode is set', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				statusCode: 201
			};

			const { statusCodeForPatternMatching } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(statusCodeForPatternMatching, '[201]');
		});

		it('Should throw an Error if statusCode is 4xx', () => {

			const originalObject = {
				statusCode: 400,
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				}
			};

			assert.throws(() => ApiResponse.send(originalObject));
		});

		it('Should throw an Error if statusCode is 5xx', () => {

			const originalObject = {
				statusCode: 500,
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				}
			};

			assert.throws(() => ApiResponse.send(originalObject));
		});

		it('Should parse the messageVariables when present to ensure that only scalar values are sent', () => {

			const originalObject = {
				statusCode: 500,
				body: {
					message: 'common.error.someMessage',
					messageVariables: {
						string: 'string',
						number: 100,
						boolean1: true,
						boolean2: false,
						array: ['foo', 100, {}],
						object: {
							foo: 'bar',
							baz: 100
						},
						null: null,
						undefined
					}
				}
			};

			const expectedError = {
				...originalObject,
				statusCodeForPatternMatching: '[500]',
				body: {
					message: 'common.error.someMessage',
					messageVariables: JSON.stringify({
						string: 'string',
						number: 100,
						boolean1: true,
						boolean2: false,
						array: 'foo,100,[object Object]',
						object: '{ foo: \'bar\', baz: 100 }',
						null: '',
						undefined: ''
					})
				}
			};

			assert.throws(() => ApiResponse.send({ ...originalObject }), err => {
				assert.deepStrictEqual(JSON.parse(err.message), expectedError);
				return true;
			});
		});

		it('Should pass the response headers', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				headers: {
					'x-foo': 'bar'
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'x-foo': 'bar'
			});
		});

		it('Should ignore cookies if it\'s an empty object', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, undefined);
		});

		it('Should set a simple cookie if it\'s defined', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {
					foo: 'bar'
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'Set-Cookie': 'foo=bar'
			});
		});

		it('Should set a simple cookie if it\'s defined, mantaining other response headers', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				headers: {
					'x-foo': 'x-bar'
				},
				cookies: {
					foo: 'bar'
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'x-foo': 'x-bar',
				'Set-Cookie': 'foo=bar'
			});
		});

		it('Should set a complex cookie if it\'s defined only with it\'s value', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {
					foo: {
						value: 'bar'
					}
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'Set-Cookie': 'foo=bar'
			});
		});

		it('Should set a complex cookie if it\'s defined', () => {

			const expireDate = new Date().toUTCString();

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {
					foo: {
						value: 'bar',
						httpOnly: true,
						secure: true,
						path: '/',
						expires: expireDate,
						domain: '.example.com'
					}
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'Set-Cookie': `foo=bar; HttpOnly; Secure; Path=/; Expires=${expireDate}; Domain=.example.com`
			});
		});

		it('Should set only the first cookie if more than one are defined', () => {

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {
					foo: 'bar',
					baz: 'yeah'
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'Set-Cookie': 'foo=bar'
			});
		});

		it('Should handle an Date object as expire date', () => {

			const expireDate = new Date();

			const originalObject = {
				body: {
					foo: 'bar',
					baz: [1, 2, 3]
				},
				cookies: {
					foo: {
						value: 'bar',
						expires: expireDate
					}
				}
			};

			const { headers } = ApiResponse.send({ ...originalObject });

			assert.deepStrictEqual(headers, {
				'Set-Cookie': `foo=bar; Expires=${expireDate.toUTCString()}`
			});
		});
	});

	describe('sendError', () => {

		it('Should set statusCode 500 if no error.statusCode is specified', () => {

			const originalError = new Error('Generic error');

			assert.throws(() => ApiResponse.sendError(originalError), e => {
				const response = JSON.parse(e.message);
				return response.statusCode === 500;
			});
		});

		it('Should set custom statusCode if error.statusCode is specified', () => {

			const originalError = new Error('Generic error');
			originalError.statusCode = 400;

			assert.throws(() => ApiResponse.sendError(originalError), e => {
				const response = JSON.parse(e.message);
				return response.statusCode === 400;
			});
		});

		it('Should use the error message as body for default', () => {

			const originalError = new Error('Generic error');

			assert.throws(() => ApiResponse.sendError(originalError), e => {
				const response = JSON.parse(e.message);
				return response.body.message === 'Generic error';
			});
		});

		it('Should set custom body if error.body is specified', () => {

			const originalError = new Error('Generic error');
			originalError.body = {
				custom: true
			};

			assert.throws(() => ApiResponse.sendError(originalError), e => {
				const response = JSON.parse(e.message);
				assert.deepStrictEqual(response.body, {
					custom: true
				});

				return true;
			});
		});

		it('Should set response headers if error.headers is specified', () => {

			const originalError = new Error('Generic error');
			originalError.headers = {
				'x-custom': 'foo'
			};

			assert.throws(() => ApiResponse.sendError(originalError), e => {
				const response = JSON.parse(e.message);
				assert.deepStrictEqual(response.headers, {
					'x-custom': 'foo'
				});

				return true;
			});
		});
	});

});


.eslintrc.js
'use strict';

module.exports = {
	extends: 'airbnb-base',

	env: {
		node: true,
		es6: true,
		mocha: true
	},

	parserOptions: {
		sourceType: 'script',
		ecmaVersion: 2020
	},

	rules: {
		'operator-linebreak': 0,
		'no-continue': 0,
		'no-plusplus': 0,
		'prefer-spread': 0,
		'prefer-rest-params': 0,
		'class-methods-use-this': 0,
		'consistent-return': 0,
		'prefer-template': 0,
		'import/extensions': 0,
		'import/no-unresolved': 0,
		'import/no-extraneous-dependencies': ['error', { devDependencies: ['**/tests/**/*.js'] }],

		'no-bitwise': 0,

		curly: ['error', 'multi-or-nest'],

		'no-underscore-dangle': ['warn', {
			allowAfterThis: true,
			allowAfterSuper: true,
			allow: ['_call', '__rootpath', '_where']
		}],

		'no-tabs': 0,

		'no-new': 0,

		'func-names': 0,

		'space-before-function-paren': ['error', {
			'anonymous': 'never',
			'named': 'never',
			'asyncArrow': 'always'
		}],

		'arrow-parens': ['error', 'as-needed'],
		'arrow-body-style': 0,

		indent: ['error', 'tab', {
			SwitchCase: 1
		}],

		'comma-dangle': ['error', 'never'],

		'padded-blocks': 0,

		'max-len': ['error', {
			code: 150,
			tabWidth: 1,
			comments: 200
		}],

		'spaced-comment': ['error', 'always', {
			exceptions: ['*']
		}],

		'newline-per-chained-call': ['error', {
			ignoreChainWithDepth: 2
		}],

		'no-param-reassign': 0,

		'no-prototype-builtins': 0,

		'keyword-spacing': ['error', {
			overrides: {
				if: {
					after: false
				},
				for: {
					after: false
				},
				while: {
					after: false
				},
				switch: {
					after: false
				},
				catch: {
					after: false
				}
			}
		}],

		'no-restricted-syntax': ['error', 'ForInStatement', 'LabeledStatement', 'WithStatement'],
		'function-paren-newline': 0,
		'no-await-in-loop': 0,

		'object-curly-newline': ['error', {
			ObjectExpression: { minProperties: 5, multiline: true, consistent: true },
			ObjectPattern: { minProperties: 5, multiline: true, consistent: true }
		}],
		'nonblock-statement-body-position': ['error', 'below', { overrides: { else: 'any' } }]
	}
};


lib/api-response.js
'use strict';

const { inspect } = require('util');
const logger = require('lllog')();

const ensureScalar = value => {

	if(typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean')
		return value;

	if(value === undefined || value === null)
		return '';

	if(Array.isArray(value))
		return value.toString();

	return inspect(value);
};

class ApiResponse {

	/**
	 * Sends a response for Api Gateway
	 *
	 * @param {object} response The response object. Properties: statusCode, body
	 * @return {object} The standarized success response object
	 * @throws {Error} When response is not successful
	 */
	static send(response) {

		const statusCode = response.statusCode || 200;

		const slsResponse = {
			...response,
			statusCode,
			statusCodeForPatternMatching: `[${statusCode}]`
		};

		this.setCookies(slsResponse);

		return this.handleResponse(slsResponse);
	}

	/**
	 * Sends an error response for Api Gateway
	 *
	 * @param {Error} error The response error. Properties: statusCode, headers, cookies, body, message, stack
	 * @param {string?} clientCode The client code of the request if present
	 * @throws {Error} Throws the error response
	 */
	static sendError(error, clientCode) {

		logger.error(clientCode, error);

		const statusCode = error.statusCode || 500;

		const slsResponse = {
			...clientCode && { clientCode },
			statusCode,
			statusCodeForPatternMatching: `[${statusCode}]`,
			...error.headers && { headers: error.headers },
			body: error.body || { message: error.message }
		};

		this.setCookies(slsResponse);

		return this.handleResponse(slsResponse);
	}

	static setCookies(response) {

		const cookieHeader = this.formatCookies(response);

		if(cookieHeader) {
			if(!response.headers)
				response.headers = {};

			response.headers['Set-Cookie'] = cookieHeader;
		}
	}

	/**
	 * Format cookies to a header value. It only formats the first cookie
	 *
	 * @param {Object} arg1 The response object
	 * @param {Object} arg1.cookies The cookies from the response object
	 * @return {string} The first cookie formatted as a header value
	 */
	static formatCookies({ cookies }) {

		if(!cookies)
			return;

		const firstCookie = Object.entries(cookies)[0];

		if(!firstCookie)
			return;

		const [cookieName, cookieValue] = firstCookie;

		if(typeof cookieValue === 'object') {

			const httpOnly = cookieValue.httpOnly ? '; HttpOnly' : '';
			const secure = cookieValue.secure ? '; Secure' : '';
			const path = cookieValue.path ? `; Path=${cookieValue.path}` : '';
			const expires = cookieValue.expires
				? (`; Expires=${cookieValue.expires.constructor.name === 'Date' ? cookieValue.expires.toUTCString() : cookieValue.expires}`)
				: '';
			const domain = cookieValue.domain ? `; Domain=${cookieValue.domain}` : '';

			return `${cookieName}=${cookieValue.value}${httpOnly}${secure}${path}${expires}${domain}`;
		}

		return `${cookieName}=${cookieValue}`;
	}

	static handleResponse(response) {

		if(response.statusCode < 400) {
			response.body = JSON.stringify(response.body);
			return response;
		}

		if(!response.body.messageVariables || typeof response.body.messageVariables !== 'object')
			throw new Error(JSON.stringify(response));

		const messageVariables = {};
		Object.entries(response.body.messageVariables).forEach(([key, value]) => {
			messageVariables[key] = ensureScalar(value);
		});

		throw new Error(JSON.stringify({
			...response,
			body: {
				...response.body,
				messageVariables: JSON.stringify(messageVariables)
			}
		}));
	}

}

module.exports = ApiResponse;
